### ISSUE TITLE
How to use refterm for something?

### ISSUE THREAD

**Issue by whisperity** [cmuratori/refterm#41]
I understand that this terminal program is only possible to use on Windows, right? But the problem is, I did get in front of my Windows machine to try it, and I am having big big **HUGE** issues...
I downloaded the `v2` release. It starts fine, it's blinking, I see the sanscrit(?) script, so it should be good.

But no matter what command I try to execute, nothing useful happens. I read and understood that this isn't a fully fledged VTE like Konsole, Gnome-Terminal, Kitty, etc. would be, but my use-case (on Windows, anyways...) is pretty straight-forward. I don't want to use any magic POSIX-y stuff and interactive stuff and ncurses, just a raw colourful dump of some textual data!

I have some huge files (***HUGE***) for which I want to render a colourful `git diff` and just look at it visually. (This is very slow in conventional terminal emulators like Konsole, Gnome-Terminal, or Contour...) But if I type `git` or `git.exe` into the prompt(?), I get that `git.exe` (or `git.exe.exe`???) is not an executable, bla bla bla. Weird, as `CMD.EXE` picks up on my Git install (from Git Bash, but even the MinGW Git Bash tends to be veeery slow for these huge files!) and produces output.

I also tried starting `CMD.EXE`. When I type something in, the original `>` prompt that blinks disappears, which means I cannot see what is the command that I typed in.

Then I get the header/MOTD/version/whatever of the `CMD.EXE` and its prompt... but then the starter `>` comes up again, and doing anything, like saying `git.exe` or `DIR` just doesn't go through. I can't `CD` to another directory as it says that the file paths or volume labels (?) cannot be found, etc.

![The visuals after executing commands `cmd` and `dir`](https://user-images.githubusercontent.com/1969470/140639735-91268b89-e97e-4c02-b7b2-a4ef5ba43530.png)

> Side track: Am I doing something bad? The letters look incredibly weird here...

It feels like `CMD.EXE` is kind of running in the background and it does print to this window (as opposed to opening up its own window!), but for some reason I cannot "communicate"(?) with the process itself?

![Process tree](https://user-images.githubusercontent.com/1969470/140639796-4773db97-4dfc-4552-8aa1-53fbb2c70d26.png)

I did develop a command-line tool for Linux to help you set your environment variables, and when rigging the automated but interactive testing infrastructure, I ran into similar issues and I needed to hack my way through attaching (from Python...) the proper std stream to the proper process... But my Windows-specific knowledge is very, very limited and I have absolutely no idea how I would go about debugging this.

Running Windows 10 on version 20.04. This is in general my gaming and fun-having machine, so I have absolutely no developer tools, such as a compiler, or WSL, or HyperV, or anything like that, installed. (The use-case mentioned above also relates to the gaming/modding part, not a hardcore developer part.)

**Comment by cmuratori** [comment]
As it says when it starts up, refterm is not a terminal. If you are not a programmer planning to use this codebase to build a terminal or test throughput/latency in a controlled way, you are not supposed to be using it at all. It is not meant for end users. As a simple example, you are clearly trying to use it not only as a terminal but as a terminal running a shell, which is two things you aren't supposed to be doing with it :)

\- Casey

**Comment by uspasojevic96** [comment]
So why is your circlejerk polluting WT issues with refterm bullshit then and saying that refterm is a VTE?

**Comment by whisperity** [comment]
@cmuratori Well, I am most certainly not a terminal developer per se, at least that isn't my day job anyhow (I'm dabbling in compiler frontends). However... I'm still a bit confused. How would I test the performance of this program, if I cannot get it to execute or do anything useful? If I have an imaginary test ecosystem of whatever purpose, how would I start it?

In your `README`, you also say this (as of 4fbb7d9): 

> TL;DR: refterm should be thought of as establishing a modern *minimum* speed at which a reasonable terminal should be expected to perform. [...]  If your terminal runs *slower* than refterm, that is not a good sign.

For my day job, I am a KDE user (yes, indeed, that's Linux and not Windows), and I am absolutely and bruttaly dissatisfied with Konsole. But how would I make a useful comparison between, let's say, refterm and Konsole? Assuming there exists a test ecosystem which *could* be compiled cross-platform? Or adapt some of the ideas in here to make that (or Gnome-Terminal or Kitty or whatever) better? At face value, they all seem slower, because they do things like executing a shell and such, whereas I can't even create a baseline comparison here, because I can't execute anything.

I still feel like I'm missing something, but I just can't wrap my head around it.

**Comment by cmuratori** [comment]
Anything you type on the refterm command line will be run directly, either by executing it (if the EXE path is completely resolved) or by running cmd.exe. So if you would like to test throughput, you want to do the former.

A simple example would be to make any program that writes data to stdout. There are test programs included in this repository that do that (splat and splat2). Run them directly in refterm. They have throughput measurement built in.

Another way would be to run termbench, which is designed to benchmark terminals. Again, to run it, all you have to do is put the full (relative or absolute) path into refterm's command line and hit return.

If you don't have termbench you can get it here: https://github.com/cmuratori/termbench

Again, all of this should have been self-explanatory so if you are confused about it, refterm is probably not a project that is relevant to you. It is _not_ an end-user utility, and is only useful to people who plan _to use the source code_ or to benchmark against terminals they are writing themselves.

\- Casey
